### PR TITLE
feat: get ENS Namehash [APE-1463]

### PR DESCRIPTION
**Pull request by Aviksaikat**
### What I did

fixes: #29 #31

### How I did it
delegated the `namehash` call to `raw_name_to_hash` of `web3.py` and also implemented manual method to calculate the ENS name hash

### How to verify it
```py
from ape_ens.utils import namehash
assert (
        namehash("").hex()
        == "0x0000000000000000000000000000000000000000000000000000000000000000"  # noqa: 501
    )

assert (
        namehash("ape.rocks.eth").hex()
        == "0x6294e43e29c5c1573554a68e6ff302fa867ab0d56b800f623c1abb77609d2b8d"
    )

```

### Checklist

- [x] Passes all linting checks (pre-commit and CI jobs)
- [x] New test cases have been added and are passing
- [x] Documentation has been updated
- [x] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

**Review by antazoey — COMMENTED** [review]
can you update the README? otherwise, this seems good to me.

File: .gitignore [hunk 1]
```diff
@@ -121,3 +121,4 @@ version.py
 *.swp
 *.swo
 
+.build/*
```
Comment (antazoey) [inline .gitignore:124]:
```suggestion
.build/*

```

File: README.md [hunk 2]
```diff
@@ -68,4 +68,22 @@ In [2]: convert("vitalik.eth", AddressType)
 Out[2]: '0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045'
 ```
 
+Get the `Ethereum Name Service` (ENS) namehash
```
Comment (antazoey) [inline README.md:71]:
```suggestion
Get the Ethereum Name Service (ENS) namehash using the `namehash` function:
```

File: README.md [hunk 3]
```diff
@@ -68,4 +68,22 @@ In [2]: convert("vitalik.eth", AddressType)
...
+
+```py
+from ape_ens.utils import namehash
+# or
+# from ape_ens.utils.namehash import namehash
+>> namehash("eth").hex()
+"0x93cdeb708b7545dc668eb9280176169d1c33cfd8ed6f04690a0bcc88a93fc4ae"
+
+>> namehash("foo.eth")
+HexBytes(
```
Comment (antazoey) [inline README.md:81]: can you put this on one line?

File: README.md [hunk 4]
```diff
@@ -68,4 +68,22 @@ In [2]: convert("vitalik.eth", AddressType)
...
+"0xde9b09fd7c5f901e23a3f19fecc54828e9c848539801e86591bd9801b019f84f")
+
+>> namehash("foo.eth").hex()
+"0x6294e43e29c5c1573554a68e6ff302fa867ab0d56b800f623c1abb77609d2b8d"
+
```
Comment (antazoey) [inline README.md:86]:
```suggestion
```

**Review by antazoey — APPROVED** [review]

File: README.md [hunk 5]
```diff
@@ -68,4 +68,20 @@ In [2]: convert("vitalik.eth", AddressType)
...
+# or
+# from ape_ens.utils.namehash import namehash
+>> namehash("eth").hex()
+"0x93cdeb708b7545dc668eb9280176169d1c33cfd8ed6f04690a0bcc88a93fc4ae"
+
+>> namehash("foo.eth")
+HexBytes("0xde9b09fd7c5f901e23a3f19fecc54828e9c848539801e86591bd9801b019f84f")
+
+>> namehash("foo.eth").hex()
+"0x6294e43e29c5c1573554a68e6ff302fa867ab0d56b800f623c1abb77609d2b8d"
```
Comment (fubuloubu) [inline README.md:84]: Why are these hashes different?

File: ape_ens/utils/namehash.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        )
+    return HexBytes(node)
+
+
+def namehash(name: str) -> HexBytes:
+    try:
+        return raw_name_to_hash(name)
+    # if anything happens in the future fallback to the manual implementation
+    except Exception as e:  # noqa
+        return manual_namehash(name)
```
Comment (fubuloubu) [inline ape_ens/utils/namehash.py:53]: This isn't the right way

File: ape_ens/utils/namehash.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import codecs
+import functools
+
+from ens.utils import raw_name_to_hash  # type: ignore
```
Comment (fubuloubu) [inline ape_ens/utils/namehash.py:4]: Would do
```suggestion
try:
    from ens.utils import raw_name_to_hash
except ImportError:
    raw_name_to_hash = None
```

File: ape_ens/utils/namehash.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+        return HexBytes(
+            compose(*(functools.partial(_sub_hash, label=label) for label in labels))(  # noqa: 501
+                node
+            )
+        )
+    return HexBytes(node)
+
+
+def namehash(name: str) -> HexBytes:
```
Comment (fubuloubu) [inline ape_ens/utils/namehash.py:48]: Then

namehash = raw_name_to_hash if raw_name_to_hash else manual_namehash

File: ape_ens/utils/namehash.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
+import codecs
+import functools
+
+try:
+    from ens.utils import raw_name_to_hash  # type: ignore
+except ImportError:
+    raw_name_to_hash = None
+from eth_utils import keccak
+from hexbytes import HexBytes
```
Comment (fubuloubu) [inline ape_ens/utils/namehash.py:9]: I think this is required to go after all the normal imports
```suggestion
from eth_utils import keccak
from hexbytes import HexBytes

try:
    from ens.utils import raw_name_to_hash  # type: ignore
except ImportError:
    raw_name_to_hash = None
```

File: ape_ens/utils/__init__.py [hunk 10]
```diff
@@ -0,0 +1,3 @@
+from .namehash import manual_namehash, namehash
+
+__all__ = ["namehash", "manual_namehash"]
```
Comment (antazoey) [inline ape_ens/utils/__init__.py:3]: **todo**: let's not expose `manual_namehash` as a public API but rather keep it internal.

To do this, rename the method and it's usage to `_manual_namehash`.

File: ape_ens/utils/namehash.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+try:
+    from ens.utils import raw_name_to_hash  # type: ignore
+except ImportError:
+    raw_name_to_hash = None
+
+# https://github.com/ethereum/EIPs/blob/master/EIPS/eip-137.md#namehash-algorithm
+
+
+def is_bytes(value):
```
Comment (antazoey) [inline ape_ens/utils/namehash.py:15]: **todo**: Use `is_bytes` from `eth_utils` rather than making your own

File: ape_ens/utils/namehash.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    return isinstance(value, (bytes, bytearray))
+
+
+def combine(f, g):
```
Comment (antazoey) [inline ape_ens/utils/namehash.py:19]: **suggestion**: Make all these methods besides `namehash` underscore-prefixed

**Comment by Aviksaikat** [comment]
mypy has a bug ?

**Review by antazoey — APPROVED** [review]

**Review by antazoey — APPROVED** [review]

**Review by antazoey — APPROVED** [review]

**Pull request merged**